### PR TITLE
Work on improving system performance in web runtime

### PR DESCRIPTION
**Pull request by hypotext**
_This is the old web-perf, pre-cleaning_

# Next steps

- [x] Compile energy
- [x] Compile gradients
- [x] Integrate line search, starting from the first simple example
- [x] Port bigger venn example and benchmark
- [x] Port bigger tree example and benchmark
- [x] Port L-BFGS
- [x] Pull `web-runtime` into this
- [x] Clean up this branch (split out autodiff lib and tests)
- [x] Phase out tensors
- [x] Port continuous map example 
- [x] Port some of geometry example
- [ ] Port cross-domain example and benchmark
- [ ] Document how to port an example
- [ ] Document the code and working examples
- [ ] Merge changes
- [ ] Test everything
- [ ] Write a high-level document of the new autodiff (incl. use, extension, ops)
- [ ] Merge this branch into `web-runtime`

**This PR is very experimental and not ready for review.**

Currently this PR includes:

* Custom reverse-mode autodiff (without Wengert tape implementation)
  * custom differentiable number type, gradient function, and some differentiable ops. All are located in the `Constraints` file
* Compiled energies and gradients with hyperparameters
* Line search

Roadmap:

- [x] Try compiling to overall objective function (as JS code), rather than interpreting the overall objective at runtime. 
- [ ] Benchmark the performance difference.

Working paths:

* Two sets with labels: `runpenrose set-theory-domain/twosets-simple.sub set-theory-domain/venn-small.sty  set-theory-domain/setTheory.dsl`
* ??: `runpenrose set-theory-domain/tree.sub set-theory-domain/venn-small.sty  set-theory-domain/setTheory.dsl`
* ??: `runpenrose set-theory-domain/tree-small2.sub set-theory-domain/venn-opt-test.sty set-theory-domain/setTheory.dsl` 
* Big set example with circles: `runpenrose set-theory-domain/tree.sub set-theory-domain/venn-opt-test.sty set-theory-domain/setTheory.dsl` 
* Big set example with tree:`runpenrose set-theory-domain/tree.sub set-theory-domain/tree.sty set-theory-domain/setTheory.dsl` (except for this:)

image

Should be working (but not yet):

- [x] Continuous map: `runpenrose set-theory-domain/continuousmap.sub set-theory-domain/continuousmap.sty set-theory-domain/setTheory.dsl`

image

- [x] Hyperbolic: `runpenrose hyperbolic-domain/hyperbolic-example.sub hyperbolic-domain/PoincareDisk.sty hyperbolic-domain/hyperbolic.dsl`

image

- [ ] Geometry (partial): 
image

[ ] Sets (optimizing thru computation): `runpenrose set-theory-domain/multisets.sub set-theory-domain/venn-comp-test.sty  set-theory-domain/setTheory.dsl`

image

- [ ] Cross-domain mesh-set example: `runpenrose mesh-set-domain/DomainInterop.sub mesh-set-domain/DomainInterop.sty mesh-set-domain/DomainInterop.dsl`

For ongoing documentation of performance, see #implementation in Slack.

---

# Description

Related issues: # (issue)

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.

# Examples with steps to reproduce them

# Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran [Haddock](https://github.com/penrose/penrose/wiki/Writing-and-generating-documentation#generating-html-documentation-site) and there were no errors when generating the HTML site
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass locally using `stack test`
- [ ] My code follows the style guidelines of this project (e.g.: no HLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

- [ ] numerical stability of compiled gradients (in particular, use of `EPS_DENOM`) -- do they remain stable on bigger examples?
- [ ] use of call stack to compile gradient/energy code -- does it scale to bigger examples, or does it blow the stack?

# Other documentation 

* To check the compiled energies/gradients WRT caching in translation, see these gists: [No caching](https://gist.github.com/hypotext/65fc2b06ec5a9e2fa6fe1e45168343e7), [With caching](https://gist.github.com/hypotext/1012c40bccdb62de64b24cd2ceaafee4)

### CODE REVIEW

File: react-renderer/src/Constraints.ts [hunk 1]
```diff
@@ -6,230 +6,230 @@ const TOL = 1e-4;
 const DEBUG_ENERGY = false;
 
 export const objDict = {
-  equal: (x: DiffVar, y: DiffVar) => squaredDifference(x, y),
```
Comment (hypotext) [inline react-renderer/src/Constraints.ts]: In this commit my IDE auto-indented the file, so the diff is misleading. Maybe split this commit so it doesn't pollute the blame history

File: react-renderer/src/Autodiff.ts [hunk 2]
```diff
@@ -0,0 +1,1451 @@
...
+  } else {
+    v.parentsGrad.push({ node: z, sensitivityNode: none });
+
+    z.childrenGrad.push({ node: v, sensitivityNode: none });
+  }
+
+  return z;
+};
+
+// TODO: rename to `square` after tf.js dependency is removed
```
Comment (wodeni) [inline react-renderer/src/Autodiff.ts:423]: Seems like tfjs is gone. Resolve now?

**Review by wodeni — APPROVED** [review]
The changes look good to me! I'm no expert in AD, so my comments are perhaps too superficial. Lmk if you have specific questions in mind.

File: react-renderer/src/Autodiff.ts [hunk 3]
```diff
@@ -0,0 +1,1451 @@
+import * as _ from "lodash";
```
[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/Autodiff.ts:1]: Let's do this during an eng meeting :)

File: react-renderer/src/Optimizer.ts [hunk 4]
```diff
@@ -1,71 +1,91 @@
...
-import { constrDict, objDict } from "./Constraints";
+import * as _ from "lodash";
+import {
+  constrDict,
+  objDict,
+} from "./Constraints";
+import { scalev, addv, subv, negv, dot, prettyPrintFns, prettyPrintProperty } from "./OtherUtils";
+
+// For deep-cloning the translation
+const clone = require('rfdc')({ proto: false, circles: true });
```
[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/Optimizer.ts:29]: It's only used in a couple places, I think it's ok here.

File: react-renderer/src/codegen-results.md [hunk 5]
```diff
@@ -0,0 +1,359 @@
+# Test 1 with codegen
```
[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/codegen-results.md:1]: Yep, the tests in Autodiff.ts are the ones that should go in the test suite

[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/codegen-results.md:1]: I just removed this file

File: react-renderer/src/Evaluator.ts [hunk 6]
```diff
@@ -1,12 +1,45 @@
...
+import { floatVal } from "./OtherUtils";
 import seedrandom from "seedrandom";
-import { Tensor, Variable, scalar } from "@tensorflow/tfjs";
-import { sc, differentiable, evalEnergyOn } from "./Optimizer";
-import { compDict, checkComp } from "./Computations";
-import { mapTranslation } from "./EngineUtils";
+
+import { compDict, checkComp } from "./Functions";
+
+import {
```
[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/Evaluator.ts:16]: Good point, I can do this later. Just want to get `web-perf` in asap

File: react-renderer/src/Protocol.tsx [hunk 7]
```diff
@@ -53,13 +53,30 @@ export class Protocol {
     };
     this.ws.onerror = this.onSocketError;
   };
+
+  public clearPacket = (packet: any) => {
```
[the remark this answers is not in the document]
Comment (hypotext) [reply react-renderer/src/Protocol.tsx:57]: Let's do this in an eng meeting :)

**Comment by hypotext** [comment]
Thanks for the review @wodeni! Gonna close this because I'm moving to a cleaner branch.

File: react-renderer/src/Evaluator.ts [hunk 8]
```diff
@@ -230,7 +281,7 @@ export const evalExpr = (
       return {
         tag: "Val",
         // HACK: coerce the type for now to let the compiler finish
-        contents: evalUOp(uOp, arg as IFloatV<Tensor> | IIntV<Tensor>),
+        contents: evalUOp(uOp, arg as IFloatV<VarAD> | IIntV<VarAD>),
       }
     } break;
```
Comment (wodeni) [inline react-renderer/src/Evaluator.ts:286]: Sorry about missing these. I don't think the `break;`s do anything here.

Comment (hypotext) [reply react-renderer/src/Evaluator.ts:286]: Forgot to comment but I did fix this earlier

**Pull request closed without merging**